File: C/Hacks/Hack10/proteinTranslator.c
/*
 * Name: Anthony Candelas
 * Email: tsallinger2@unl.edu
 * Date: 2025-11-10
 * Lab/Task: Hack10
 *
 * Translates a DNA sequence (from an input file) into a protein
 * sequence and writes the result to an output file. The program
 * ignores whitespace in the input, converts T->U to form RNA, then
 * translates codons using the provided rnaToProtein function.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#include "file_utils.h"
#include "protein_utils.h"

int main(int argc, char **argv) {
  if (argc != 3) {
    return 1;
  }

  const char *inPath = argv[1];
  const char *outPath = argv[2];

  char *contents = getFileContents(inPath);
  if (contents == NULL) {
    return 1;
  }

  /* Build RNA string by removing whitespace and converting T->U. */
  size_t len = strlen(contents);
  char *rna = (char *)malloc(len + 1);
  if (rna == NULL) { free(contents); return 1; }
  size_t rpos = 0;
  for (size_t i = 0; i < len; i++) {
    char c = contents[i];
    if (isspace((unsigned char)c)) continue;
    char up = (char)toupper((unsigned char)c);
    if (up == 'T') up = 'U';
    /* we only keep A, C, G, U characters */
    if (up == 'A' || up == 'C' || up == 'G' || up == 'U') {
      rna[rpos++] = up;
    } else {
      /* invalid character in DNA -- treat as error */
      free(contents);
      free(rna);
      return 1;
    }
  }
  rna[rpos] = '\0';

  /* Translate RNA into protein, codon-by-codon */
  size_t protein_cap = (rpos / 3) + 1;
  char *protein = (char *)malloc(protein_cap + 1);
  if (protein == NULL) { free(contents); free(rna); return 1; }
  size_t ppos = 0;

  for (size_t i = 0; i + 2 < rpos; i += 3) {
    char codon[4] = { rna[i], rna[i+1], rna[i+2], '\0' };
    char aa = rnaToProtein(codon);
    if (aa == 0) {
      free(contents);
      free(rna);
      free(protein);
      return 1;
    }
    if (aa == 'x') break; /* stop codon encountered */
    protein[ppos++] = aa;
  }
  protein[ppos] = '\0';

  /* Write protein to output file */
  FILE *out = fopen(outPath, "w");
  if (out == NULL) {
    free(contents);
    free(rna);
    free(protein);
    return 1;
  }
  fprintf(out, "%s\n", protein);
  fclose(out);

  free(contents);
  free(rna);
  free(protein);
  return 0;
}

File: Java/Hack10/unl/soc/ProteinTranslator.java
package unl.soc;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Locale;

/**
 * Name: Anthony Candelas
 * Email: tsallinger2@unl.edu
 * Date: 2025-11-10
 * Lab/Task: Hack10
 *
 * Command-line wrapper that reads a DNA file, converts to RNA,
 * translates to protein using ProteinTranslator.proteinMap, and
 * writes the protein string to the output file. On any error the
 * program exits with a non-zero status (no extra prints).
 */
public class ProteinTranslator {
  /**
   * Trigram RNA to protein map. Lower-case 'x' marks stop codons.
   */
  public static final java.util.Map<String, Character> proteinMap = new java.util.HashMap<>();

  static {
    proteinMap.put("AAA", 'K');
    proteinMap.put("AAC", 'N');
    proteinMap.put("AAG", 'K');
    proteinMap.put("AAU", 'N');
    proteinMap.put("ACA", 'T');
    proteinMap.put("ACC", 'T');
    proteinMap.put("ACG", 'T');
    proteinMap.put("ACU", 'T');
    proteinMap.put("AGA", 'R');
    proteinMap.put("AGC", 'S');
    proteinMap.put("AGG", 'R');
    proteinMap.put("AGU", 'S');
    proteinMap.put("AUA", 'I');
    proteinMap.put("AUC", 'I');
    proteinMap.put("AUG", 'M');
    proteinMap.put("AUU", 'I');
    proteinMap.put("CAA", 'Q');
    proteinMap.put("CAC", 'H');
    proteinMap.put("CAG", 'Q');
    proteinMap.put("CAU", 'H');
    proteinMap.put("CCA", 'P');
    proteinMap.put("CCC", 'P');
    proteinMap.put("CCG", 'P');
    proteinMap.put("CCU", 'P');
    proteinMap.put("CGA", 'R');
    proteinMap.put("CGC", 'R');
    proteinMap.put("CGG", 'R');
    proteinMap.put("CGU", 'R');
    proteinMap.put("CUA", 'L');
    proteinMap.put("CUC", 'L');
    proteinMap.put("CUG", 'L');
    proteinMap.put("CUU", 'L');
    proteinMap.put("GAA", 'E');
    proteinMap.put("GAC", 'D');
    proteinMap.put("GAG", 'E');
    proteinMap.put("GAU", 'D');
    proteinMap.put("GCA", 'A');
    proteinMap.put("GCC", 'A');
    proteinMap.put("GCG", 'A');
    proteinMap.put("GCU", 'A');
    proteinMap.put("GGA", 'G');
    proteinMap.put("GGC", 'G');
    proteinMap.put("GGG", 'G');
    proteinMap.put("GGU", 'G');
    proteinMap.put("GUA", 'V');
    proteinMap.put("GUC", 'V');
    proteinMap.put("GUG", 'V');
    proteinMap.put("GUU", 'V');
    proteinMap.put("UAA", 'x');
    proteinMap.put("UAC", 'Y');
    proteinMap.put("UAG", 'x');
    proteinMap.put("UAU", 'Y');
    proteinMap.put("UCA", 'S');
    proteinMap.put("UCC", 'S');
    proteinMap.put("UCG", 'S');
    proteinMap.put("UCU", 'S');
    proteinMap.put("UGA", 'x');
    proteinMap.put("UGC", 'C');
    proteinMap.put("UGG", 'W');
    proteinMap.put("UGU", 'C');
    proteinMap.put("UUA", 'L');
    proteinMap.put("UUC", 'F');
    proteinMap.put("UUG", 'L');
    proteinMap.put("UUU", 'F');
  }

  public static void main(String[] args) {
    if (args == null || args.length != 2) {
      System.exit(1);
    }
    String inPath = args[0];
    String outPath = args[1];

    String contents = FileUtils.getFileContents(inPath);
    if (contents == null) System.exit(1);

    StringBuilder rna = new StringBuilder(contents.length());
    for (int i = 0; i < contents.length(); i++) {
      char c = contents.charAt(i);
      if (Character.isWhitespace(c)) continue;
      char up = Character.toUpperCase(c);
      if (up == 'T') up = 'U';
      if (up == 'A' || up == 'C' || up == 'G' || up == 'U') {
        rna.append(up);
      } else {
        System.exit(1);
      }
    }

    StringBuilder protein = new StringBuilder(rna.length() / 3 + 1);
    for (int i = 0; i + 2 < rna.length(); i += 3) {
      String codon = rna.substring(i, i+3).toUpperCase(Locale.ROOT);
      Character aa = ProteinTranslator.proteinMap.get(codon);
      if (aa == null) System.exit(1);
      if (aa == 'x') break;
      protein.append(aa.charValue());
    }

    try {
      Files.write(Paths.get(outPath), (protein.toString() + System.lineSeparator()).getBytes(StandardCharsets.UTF_8));
    } catch (IOException e) {
      System.exit(1);
    }
    System.exit(0);
  }

}
